### ISSUE TITLE
[API Proposal]: Decouple ValidateOnStart from Hosting

### ISSUE THREAD

**Issue by tekian** [dotnet/runtime#84347]
Usage of [ValidateOnStart()](https://learn.microsoft.com/dotnet/api/microsoft.extensions.dependencyinjection.optionsbuilderextensions.validateonstart?view=dotnet-plat-ext-7.0) today is located in the `Microsoft.Extensions.Hosting` assembly\package. This is problematic as it greatly inflates dependency tree of packages that need to validate their options eagerly when the hosting package is not otherwise needed. The `ValidateOnStart()` functionality should be moved to the `Microsoft.Extensions.Options` package.

In order for a host to call a "Validate()" method, a new public interface needs to be added, called `IValidateOnStart` here. This interface will be injected by the `ValidateOnStart()` functionality and then obtained by a host (for the default host, in the  `Microsoft.Extensions.Hosting` assembly) with a call to `GetService<IValidateOnStart>()` in order to trigger the validation during startup.

The new interface located in the `Microsoft.Extensions.Options` assembly:
```diff
namespace Microsoft.Extensions.Options
{
+    public interface IValidateOnStart
+    {
+        void Validate();
+   }
}
```

The existing `ValidateOnStart()` extension method is located on the `Microsoft.Extensions.DependencyInjection.OptionsBuilderExtensions` extension class in the `Microsoft.Extensions.Hosting` assembly. The `Microsoft.Extensions.Hosting` assembly needs to forward this to the `Microsoft.Extensions.Options` assembly:
```diff
+[assembly: System.Runtime.CompilerServices.TypeForwardedTo(
+    typeof(Microsoft.Extensions.DependencyInjection.OptionsBuilderExtensions))]
```

### Background and motivation

Usage of `ValidateOnStart` today is tightly coupled with Microsoft.Extensions.Hosting package. The call registers both `ValidationHostedService` and instructs `ValidateOnStart` infrastructure to validate `T`-type upon startup. This is problematic as it greatly inflates dependency tree of packages that need to validate their options eagerly.

Following measures should be considered:
- Move `ValidateOnStart` to Microsoft.Extensions.Options package. This decouple registration from implementation. Code that expresses the desire for eager validation doesn't involve registration of ValidationHostedService in the same call. The new API enables a cleaner separation of concern.
- Incorporate `ValidationHostedService` into `Host` bootstrapping as a shared option infrastructure. It becomes part of default `Host` and doesn't need to be registered with every `ValidateOnStart` call.

### API Proposal

```csharp
namespace Microsoft.Extensions.Options
{
    public static class OptionsBuilderExtensions
    {
        public static OptionsBuilder<TOptions> ValidateOnStart<TOptions>(this OptionsBuilder<TOptions> builder) 
            where TOptions : class;
    }
```

(Shape of the API remains the same)

### API Usage

(Usage of the API remains the same.)

### Risks

Moving types around.

**Comment by eerhardt** [comment]
See https://github.com/dotnet/runtime/pull/47821#discussion_r573383024 for some reasons why it was originally done this way.

Maybe Hosting needs an `IStartupService` enumerable service that gets resolved and all of them notified on start of the Host. My suggestion in the above conversation was a hook in the DI container (to know when the container was built), but that was rejected because we can't rely on the specifics of our DI container implementation. It would have to be a feature across all DI implementations for us to take advantage of it.

**Comment by tekian** [comment]
@eerhardt Thanks for the context. What would be the difference between `IStartupService` and `IHostedService`? Regardless of which of the two it would use, I believe "the desire to validate options eagerly" can be safely de-coupled from "implementation that performs it". We could have `AddValidateOnStart(this IServiceCollection...` inside Microsoft.Extensions.Hosting package that would add the service. And `ValidateOnStart(this IOptionsBuilder...` would be in Microsoft.Extensions.Options.

**Comment by eerhardt** [comment]
> What would be the difference between IStartupService and IHostedService? 

The difference would be that it wouldn’t need to be kept alive for the entire process just to call a no-op Stop method on it during shutdown. It would just get created, notified, and disposed during Start.

**Comment by ericstj** [comment]
@tekian can you help me understand the problem statement here?  I see
> This is problematic as it greatly inflates dependency tree of packages that need to validate their options eagerly.

Does this mean you are able to accomplish your goals today but you end up depending on more packages than you'd like?  How does this correspond to the customer experience?

**Comment by tekian** [comment]
@ericstj The idea is to decouple the options validation from the hosting layer, allowing **libraries** (not only services) to depend only on `Microsoft.Extensions.Options` instead of `Microsoft.Extensions.Hosting`. `Microsoft.Extensions.Hosting` has a significant dependency footprint.

I propose to decouple `ValidationHostedService` from `ValidateOnStart`, even introduce a separate provider for eager options validation that can be registered and configured during the Host buildout. This way, consumers could replace it with their own implementations if desired.

**Comment by ericstj** [comment]
If a library needs hosting anyways to do anything useful in an application what do you accomplish by trying to remove the dependency?

Are you trying to say you want to support someone doing eager validation without using hosting at all?  If so that wouldn’t need more API than what is proposed here.  Where is the API that the user would trigger to indicate Start?

**Comment by ericstj** [comment]
@davidfowl @eerhardt what do you think about this separation of responsibility?  It allows libraries to opt in their types to eager validation without depending on hosting.

One challenge this still has is that it would still require someone to register that `ValidationHostedService`.  That could be avoided by making the host resolve the registered singleton (either `IValidateOnStart` or `IOptions<ValidateOnStartOptions>`) and call Validate, rather than adding the `ValidationHostedService`.

**Comment by davidfowl** [comment]
This sounds fine to me. It's less about validating on start and more about exposing an API that does the validation decoupled from hosting. Anyone can call it and hosting can too.

**Comment by eerhardt** [comment]
I like the goal and think it is worthwhile to pursue a solution to it.

I think there is a more general pattern here between this issue and https://github.com/dotnet/runtime/issues/43149. The pattern I see is "I want to run some code when the service container is built". There is also https://github.com/dotnet/runtime/issues/84847, which is very similar except it is asking for a hook just before the container is built.

I question whether this should be at the `Hosting` layer or built into `DependencyInjection`. Note there are other app models like Blazor WASM and MAUI that don't use Microsoft.Extensions.Hosting. MAUI has a similar concept here: https://learn.microsoft.com/en-us/dotnet/api/microsoft.maui.hosting.imauiinitializeservice.initialize, which is retrieved and notified when the application is built at startup time: 

https://github.com/dotnet/maui/blob/b18e80609dd00957cafcc886bf76743cf32e2759/src/Core/src/Hosting/MauiAppBuilder.cs#L149-L165

@DamianEdwards was following a similar pattern in https://github.com/aspnet/Benchmarks/pull/1836/files#r1175547227 using an IHostedService just to get a Database initializer executed when the app starts up.

In summary, I think we need a general mechanism for running code "at startup". Doing Option validation is just one of those things that needs this.

**Comment by rafal-mz** [comment]
I think we not necessary need a new abstraction for that. The list of `IHostedService` is already executed sequentually at the startup. As far as I know, the `BackgroundService` is run in the background, and `IHostedService` should finish rather sooner than later. 

I think we could have a slot for first hosted service in the list:

Something like:

```csharp
   if (services.Count != 0 && services[0].ImplementationType == typeof(StartupHostedService))
  {
      return;
  }

  services
      .RemoveAll<StartupHostedService>()
      .Insert(0, ServiceDescriptor.Singleton<IHostedService, StartupHostedService>());
```

The startup hosted service then, could accept a list of some jobs that are executed independently in concurrent fashion. For instance with the following signature:

```csharp
/// <summary>
/// Holds the initialization function, so we can pass it through <see cref="IServiceProvider"/>.
/// </summary>
public interface IStartupInitializer
{
    /// <summary>
    /// Short startup initialization job.
    /// </summary>
    /// <param name="token">Cancellation token.</param>
    /// <returns>New <see cref="Task"/>.</returns>
    Task InitializeAsync(CancellationToken token);
}
```

To register such function, you would provide a timeout for such job. You could also use a function not to introduce an interface for simple cases:

```csharp
AddStartupInitializer(this IServiceCollection services, Func<IServiceProvider, CancellationToken, Task> initializer);

AddStartupInitializer<T>()  where T : class, IStartupInitializer;
```

We could also consider adding a name for functions, so that there are no duplicate initializers.
I think such model fits nicely with the options validation model, populating database with the seed, checking DNS records at startup and so on.

**Comment by ericstj** [comment]
So it sounds like folks are more/less OK with exposing a Startup method from options and we have some questions about how that might be plugged into Hosting. 

@rafal-mz it just so happens we were already planning on adding a version of `IStartupInitializer` to hosting.  It makes sense to have this feature use that functionality for registration.  Probably we need to think about where `IStartupInitializer` lives and what is the gesture to register them.  Do we require some explicit API call, or is it sufficient to just have one in the service collection?

If we put `IStartupInitializer` low enough and made consumption automatic in hosting, then we could pull off the move of the `ValidateOnStart` functionality without even adding any public API from Options (though we still could do that if we wanted to) or requiring explicit registration (though we could still do that if we wanted to).  Perhaps this was what @tekian was implying the whole time but I was being too dense to follow.  Regardless I'm with you now 😄.

I suppose the question is now -- do we split off the `IStartupInitializer` into a separate issue or just include it here?  cc @steveharter who's starting to look at #43149.

**Comment by joperezr** [comment]
@ericstj just making sure that the ball keeps on rolling here, from your previous message I'm not sure if you think that this is ready for review or if there is something we still want to validate via either @steveharter and/or @rafal-mz. Would you mind specifying what are the next steps here?

**Comment by ericstj** [comment]
@steveharter will be drafting an API proposal for `IStartupInitializer`-functionality.  Once this is done the work for Options (this issue) and DI (https://github.com/dotnet/runtime/issues/43149) can proceed and depend on `IStartupInitializer`.

**Comment by bartonjs** [comment]
We changed the interface name from `IValidateOnStart` (describing when it's called) to `IStartupValidator` (describing what it is).

```C#
namespace Microsoft.Extensions.Options
{
    public interface IStartupValidator
   {
        void Validate();
   }
}
```

and we acknowledge the movement of OptionsBuilderExtensions

```diff
+[assembly: System.Runtime.CompilerServices.TypeForwardedTo(
+    typeof(Microsoft.Extensions.DependencyInjection.OptionsBuilderExtensions))]
```

**Comment by joperezr** [comment]
@tekian @geeknoid looks like you guys weren't present on the above discussion. Any concerns with the decisions made above?

**Comment by joperezr** [comment]
PR implementaiton for this is out: https://github.com/dotnet/runtime/pull/88546.

Driven by @steveharter and the target is to merge by P7 code-complete date (7/14).
